File: 2035-count-sub-islands/2035-count-sub-islands.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int countSubIslands(int[][] grid1, int[][] grid2) {
        boolean[][] islands = new boolean[grid1.length][grid1[0].length];

        // Mark islands in grid1
        for (int i = 0; i < grid1.length; i++) {
            for (int j = 0; j < grid1[0].length; j++) {
                if (grid1[i][j] == 1 && !islands[i][j]) {
                    helper(grid1, i, j, islands);
                }
            }
        }

        // Traverse grid2 and count sub-islands
        boolean[][] visited = new boolean[grid2.length][grid2[0].length];
        int count = 0;
        for (int i = 0; i < grid2.length; i++) {
            for (int j = 0; j < grid2[0].length; j++) {
                if (grid2[i][j] == 1 && !visited[i][j] && islands[i][j]) {
                    count += helper2(grid2, i, j, islands, visited);
                }
            }
        }

        return count;
    }

    private void helper(int[][] grid, int row, int col, boolean[][] visited) {
        Queue<int[]> q = new LinkedList<>();
        int[][] dirs = { { 1, 0 }, { 0, 1 }, { -1, 0 }, { 0, -1 } };
        q.offer(new int[] { row, col });
        visited[row][col] = true;

        while (!q.isEmpty()) {
            int[] cell = q.poll();
            for (int[] dir : dirs) {
                int newR = cell[0] + dir[0];
                int newC = cell[1] + dir[1];

                if (newR >= 0 && newR < grid.length && newC >= 0 &&
                        newC < grid[0].length && grid[newR][newC] == 1 && !visited[newR][newC]) {
                    q.offer(new int[] { newR, newC });
                    visited[newR][newC] = true;
                }
            }
        }
    }

    private int helper2(int[][] grid, int row, int col, boolean[][] islands, boolean[][] visited) {
        Queue<int[]> q = new LinkedList<>();
        int[][] dirs = { { 1, 0 }, { 0, 1 }, { -1, 0 }, { 0, -1 } };
        q.offer(new int[] { row, col });
        visited[row][col] = true;
        boolean isSubIsland = true;

        while (!q.isEmpty()) {
            int[] cell = q.poll();
            for (int[] dir : dirs) {
                int newR = cell[0] + dir[0];
                int newC = cell[1] + dir[1];

                if (newR >= 0 && newR < grid.length && newC >= 0 &&
                        newC < grid[0].length && grid[newR][newC] == 1) {

                    if (!visited[newR][newC]) {
                        if (!islands[newR][newC]) {
                            isSubIsland = false;
                        }
                        q.offer(new int[] { newR, newC });
                        visited[newR][newC] = true;
                    }
                }
            }
        }

        return isSubIsland ? 1 : 0;
    }
}

File: 2035-count-sub-islands/2035-count-sub-islands.py
class Solution(object):
    def countSubIslands(self, grid1, grid2):
        """
        :type grid1: List[List[int]]
        :type grid2: List[List[int]]
        :rtype: int
        """
        if not grid1 or not grid2:
            return 0

        m, n = len(grid2), len(grid2[0])
        visited = [[False] * n for _ in range(m)]
        count = 0

        def bfs(row, col):
            queue = deque([(row, col)])
            visited[row][col] = True
            is_sub_island = True
            directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]

            while queue:
                r, c = queue.popleft()
                
                for dr, dc in directions:
                    new_r, new_c = r + dr, c + dc

                    if 0 <= new_r < m and 0 <= new_c < n and grid2[new_r][new_c] == 1 and not visited[new_r][new_c]:
                        if grid1[new_r][new_c] != 1:
                            is_sub_island = False
                        queue.append((new_r, new_c))
                        visited[new_r][new_c] = True
            
            return 1 if is_sub_island else 0

        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 1 and grid1[i][j] == 1 and not visited[i][j]:
                    count += bfs(i, j)

        return count
